### PR TITLE
fix(gatsby-source-filesystem): Retry stalled remote file downloads

### PR DESCRIPTION
**Pull request by ascorbic**
## Description

gatsby-source-filesystem has a problem where downloads would stall and never timeout or error. Eventually all the concurrent downloads were stalled, preventing any more starting. This was causing builds to stall indefinitely. 

This PR adds a timeout in `createRemoteFileNode`. If no `data` events are received in the download stream for 30 seconds then the download is cancelled and retried up to three times. In tests this unblocks the downloads on the first attempt.

WIP while I'm writing unit tests.

### CODE REVIEW

File: packages/gatsby-source-filesystem/package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "file-type": "^12.4.0",
     "fs-extra": "^8.1.0",
     "gatsby-core-utils": "^1.0.26",
-    "got": "^7.1.0",
+    "got": "^9.6.0",
```
Comment (pieh) [inline packages/gatsby-source-filesystem/package.json]: We might need to wait until we trully drop Node 8 - we added deprecation note in https://github.com/gatsbyjs/gatsby/pull/20466, so we need to keep tracking node version usage

Comment (ascorbic) [reply packages/gatsby-source-filesystem/package.json]: Aha. In that case we could move to got@10, which drops 8 support. I *think* we can make do for now with got@8, which has the progress events that we need. I'll take a look on Monday

Comment (pieh) [reply packages/gatsby-source-filesystem/package.json]: Ah, yeah - was just writing about `got@8` ;)

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 2]
```diff
@@ -124,31 +126,65 @@ async function pushToQueue(task, cb) {
...
+
+    // Called if we stall for 30s without receiving any data
+    const handleTimeout = () => {
+      fsWriteStream.close()
+      fs.removeSync(tmpFilename)
+      if (attempt < RETRY_LIMIT) {
+        // Retry by calling ourself recursively
+        resolve(
+          requestRemoteNode(url, headers, tmpFilename, httpOpts, attempt + 1)
+        )
```
Comment (pieh) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: should we try to re-implement exponential backoff before re-trying? or the code as-is work well enough?

Comment (ascorbic) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: So far it's never needed to retry more than once. I think backoff would make sense if we were retrying after catching errors, but these stalls seem to be totally random and can be retried immeditely

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 3]
```diff
@@ -124,31 +126,65 @@ async function pushToQueue(task, cb) {
...
+    const resetTimeout = () => {
+      if (timeout) {
+        clearTimeout(timeout)
+      }
+      timeout = setTimeout(handleTimeout, STALL_TIMEOUT)
+    }
     const responseStream = got.stream(url, {
       headers,
-      ...opts,
+      timeout: 30000,
```
Comment (sidharthachatterjee) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: Want to move up to a named constant?

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 4]
```diff
@@ -124,31 +126,65 @@ async function pushToQueue(task, cb) {
...
+    }
+
     const responseStream = got.stream(url, {
       headers,
-      ...opts,
+      ...httpOpts,
     })
     const fsWriteStream = fs.createWriteStream(tmpFilename)
     responseStream.pipe(fsWriteStream)
-    responseStream.on(`downloadProgress`, pro => console.log(pro))
```
Comment (muescha) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: the `console.log` is removed accidentally resetTimeout?

**Comment by mjmaurer** [comment]
I think this is breaking my build (at least locally with `gatsby develop`).
________________________
Don't have time to grok the code, but what I think is happening:

I'm using `gatsby-source-contentful` with `downloadLocal` enabled. When `createRemoteFileNode` is called [here](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/src/download-contentful-assets.js#L72), the entire batch of images is being given the same timeout (30s in this PR). I'm guessing the timeout is somehow not being cleared.
________________________
Ultimately, what I'm sure of is that increasing the timeout to 20 min fixes my issues.

**Pull request merged**